### PR TITLE
v4: Replace hand-rolled parser with tree-sitter-bash AST

### LINKED ISSUE #7: v4: Replace hand-rolled parser with tree-sitter-bash AST
## Motivation

The current shell parser is hand-rolled character-by-character parsing for heredocs, pipes, compound operators, substitutions, quoting, and redirection. While well-tested (357 tests), each new shell feature adds complexity to a parser that isn't built on formal grammar. We've already had security-critical bugs from this approach:

- **Heredoc pipe-swallowing** (`c555f0c`): `skip_heredoc()` consumed pipe operators on the `<<DELIM` line, causing `cat <<'EOF' | kubectl apply -f -` to bypass the gate entirely
- **False redirection detection** (`3daf79c`): `>` inside heredoc bodies (e.g. email addresses in Co-Authored-By lines) triggered false ASK escalation
- **False substitution detection** (`0a60163`): Backticks in heredoc content triggered false command-substitution extraction

Each fix required understanding increasingly subtle shell syntax edge cases. The set of edge cases is effectively infinite — we're reinventing the shell parser in an ad-hoc way.

## Proposal

Replace the custom parser with [tree-sitter-bash](https://github.com/tree-sitter/tree-sitter-bash), which provides a proper AST parsed from the same grammar the editor ecosystem uses.

### What we need from the AST

We don't need to execute like a shell — we need to **decompose** like one:
1. **Pipeline segments**: What commands does this string run?
2. **Heredoc boundaries**: Where do heredoc bodies start and end?
3. **Redirections**: Does anything redirect output to a file?
4. **Command substitutions**: What's inside `$(...)` and backticks?
5. **Compound operators**: `&&`, `||`, `;`, `|`

tree-sitter-bash provides all of these as structured AST nodes.

### Tradeoffs

- **Binary size**: ~1MB increase from tree-sitter C runtime. Acceptable.
- **Dependency**: tree-sitter is a well-maintained C library with broad ecosystem usage.
- **Rewrite scope**: The evaluation layer (`eval/mod.rs`) is mostly decoupled from parsing — it calls `parse_with_substitutions()` and gets back `ParsedPipeline`. The rewrite is contained to `src/parse/`.
- **Correctness**: Bugs like heredoc pipe-swallowing become impossible by construction — the grammar handles them, not our code.

### Implementation sketch

1. Add `tree-sitter` and `tree-sitter-bash` to dependencies
2. Replace `parse_with_substitutions()` with AST-based extraction
3. Walk the AST to produce the same `ParsedPipeline` / `ShellSegment` output
4. Remove `split_compound_command()`, `extract_substitutions()`, `skip_heredoc()`, `has_output_redirection()` — all replaced by AST traversal
5. Keep all existing integration tests as regression suite (they test the evaluation layer, not parser internals)

### Success criteria

- All existing integration tests pass
- Parser-level unit tests replaced with AST-based equivalents
- No hand-rolled character-by-character shell syntax parsing remains
- The entire class of "parser missed a shell syntax edge case" bugs is eliminated

## References

- tree-sitter-bash: https://github.com/tree-sitter/tree-sitter-bash
- tree-sitter Rust bindings: https://crates.io/crates/tree-sitter
- Heredoc pipe-swallowing fix: c555f0c
- False redirection fix: 3daf79c
- False substitution fix: 0a60163

### LINKED ISSUE #10: For-loop body not evaluated through command rules (trivial-case optimization bug)
## Summary

The trivial-case optimization in `src/eval/mod.rs:214` short-circuits compound evaluation when the parser produces exactly 1 segment with no substitutions:

```rust
if pipeline.segments.len() <= 1 && substitutions.is_empty() {
    return self.evaluate_single(command); // evaluates ORIGINAL string
}
```

When `walk_loop` extracts a for-loop body (e.g. `rm "$i"` from `for i in *; do rm "$i"; done`), the pipeline has exactly 1 segment. The optimization fires and passes the **original** command string to `evaluate_single`, which tokenizes `for` as the base command → "unrecognized command: for" → ASK.

## Impact

- `for i in *; do rm "$i"; done` → ASK (correct decision, wrong reason — should cite `rm` rules, not "unrecognized: for")
- `for i in *; do ls "$i"; done` → ASK (**false positive** — should be ALLOW since `ls` is safe)
- Same pattern affects `while` loops

## Reproduction

```bash
echo '{"tool_name":"Bash","tool_input":{"command":"for i in *; do ls \"$i\"; done"}}' \
  | cargo run --release 2>/dev/null | python3 -m json.tool
```

Returns `"permissionDecision": "ask"` with `"permissionDecisionReason": "unrecognized command: for"`.

## Fix

The trivial-case check should verify the segment text matches the original command, not just count segments:

```rust
if pipeline.segments.len() <= 1 && substitutions.is_empty() {
    let seg_text = pipeline.segments.first().map(|s| s.command.trim());
    if seg_text == Some(command.trim()) || seg_text.is_none() {
        return self.evaluate_single(command);
    }
}
```

When the segment differs from the original (loop body extracted), fall through to compound evaluation which correctly applies rules to the extracted segment.

## Context

This was flagged by Codex in PR #8 review (comment ID 2825635066). Our initial response incorrectly claimed a byte-range check prevented this — no such check exists.

### PR DESCRIPTION
**Pull request by butterflysky-ai**
## Summary

Rewrites `src/parse/shell.rs` to use tree-sitter-bash for proper AST-based parsing, eliminating an entire class of parser edge-case bugs by construction. Also includes project cleanup, test restructuring, thorough rustdoc, and module reorganization.

### Parser rewrite — bugs fixed
- Shell keywords (`for`/`do`/`done`/`if`/`then`/`fi`) no longer parsed as unrecognized commands (~65 false ASK prompts eliminated)
- `2>/dev/null` no longer falsely flagged as dangerous output redirection
- Heredoc body content no longer leaks into command parsing
- Control flow constructs properly unwrapped — inner commands evaluated, keywords are grammar structure
- Heredoc same-line operators (`|`, `&&`, `||`, `;`) correctly handled via AST children
- Case statement pattern labels no longer leak as executable segments
- For/while/if loop bodies correctly evaluated against rules (trivial-case passthrough fix)

### Project cleanup
- **Test restructuring** — moved 219 integration tests from `main.rs` to `tests/integration.rs`, introduced `decision_test!` macro reducing ~188 one-liner tests to single macro invocations. Added 6 control flow integration tests.
- **README revamp** — Mermaid diagrams (decision flow, module architecture, AST walk dispatch), tree-sitter architecture section, testing guide, contributing guide.
- **Thorough rustdoc** — doc comments on all public and private structs, fields, enums, methods, and modules across 16 files. Zero \`cargo doc --document-private-items\` warnings.
- **Module restructuring** — moved tool-specific command specs (git, cargo, kubectl, gh) into \`commands/tools/\` to distinguish from generic \`SimpleCommandSpec\`. Absorbed redundant \`DenyCommandSpec\` into \`SimpleCommandSpec::new(Decision::Deny)\`.
- **Naming** — renamed \"worst decision wins\" to \"strictest decision wins\" for readability.

### Commits
- \`29f4bb0\` — v4: Replace hand-rolled parser with tree-sitter-bash AST
- \`51e6b48\` — Bump version to 0.4.0
- \`bcaabbb\` — Fix case statement pattern labels leaking as executable segments
- \`ea8606b\` — Move integration tests to tests/, add decision_test! macro, revamp README
- \`98b8cb7\` — Add thorough rustdoc across all modules
- \`8f74b89\` — Restructure commands module: tools/ hierarchy, absorb deny into simple

### Design decisions
- Same public API signatures (\`parse_with_substitutions\`, \`has_output_redirection\`) — eval layer unchanged
- Thread-local \`Parser\` (tree-sitter \`Parser\` is \`!Send\`)
- \`__SUBST__\` placeholder mechanism preserved for substitution extraction
- Conservative fallback: unknown AST nodes become single segments, eval flags as unrecognized, ASK
- Trivial-case optimization checks segment text matches original command before short-circuiting

## Test plan

- [x] 343 tests pass (118 unit + 225 integration)
- [x] \`cargo clippy\` — zero warnings
- [x] \`cargo doc --document-private-items\` — zero warnings
- [x] \`cargo build --release\` — binary builds cleanly
- [x] \`cargo nextest run\` — all pass with 2s slow-timeout detection
- [x] Manual smoke test: install release binary as hook, verify real Claude Code session

Closes #7, closes #10

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by butterflysky** [comment]
@codex review

File: src/parse/shell.rs [hunk 1]
```diff
@@ -1,740 +1,776 @@
...
-    while i < len && !chars[i].is_whitespace() && chars[i] != ';' && chars[i] != '|' && chars[i] != '&' {
-        target.push(chars[i]);
-        i += 1;
+/// `case_statement`: recurse into each `case_item`.
+fn walk_case(node: Node, source: &[u8]) -> WalkResult {
+    let mut result = WalkResult::empty();
+    let mut cursor = node.walk();
+    for child in node.named_children(&mut cursor) {
+        if child.kind() == "case_item" {
+            result.append(walk_block(child, source), Some(Operator::Semi));
```
[the remark this answers is not in the document]
Comment (butterflysky-ai) [reply src/parse/shell.rs]: Valid catch. Pattern labels like `rm)` and `kubectl)` in case statements were leaking as segments because `walk_block` recurses into all named children of `case_item`, including the pattern words before `)`. Fixed in bcaabbb with a dedicated `walk_case_item` that skips all children before the `)` delimiter and only walks the body commands after it. Added test `case_pattern_not_treated_as_command`.

File: src/parse/shell.rs [hunk 2]
```diff
@@ -1,740 +1,776 @@
...
+/// `c_style_for`, non-`do_group` children are variable names and word lists,
+/// not commands.
+fn walk_loop(node: Node, source: &[u8]) -> WalkResult {
+    let mut result = WalkResult::empty();
+    let mut cursor = node.walk();
+    for child in node.named_children(&mut cursor) {
+        match child.kind() {
+            "do_group" => result.append(walk_block(child, source), Some(Operator::Semi)),
+            _ if node.kind() == "while_statement" || node.kind() == "until_statement" => {
+                result.append(walk_ast(child, source), Some(Operator::Semi));
```
[the remark this answers is not in the document]
Comment (butterflysky) [reply src/parse/shell.rs]: @codex explain your understanding of this commented code segment and your reasoning, factoring in @butterflysky-ai's response

[the remark this answers is not in the document]
Comment (butterflysky-ai) [reply src/parse/shell.rs]: Correction: my earlier response was wrong. There is no byte-range check — the trivial-case optimization at `eval/mod.rs:214` only checks `segments.len() <= 1`, and when the for-loop body is the sole segment, it passes the **original** command string to `evaluate_single`, which tokenizes `for` as the base command → "unrecognized command: for" → ASK.

This means:
- `for i in *; do rm "$i"; done` → ASK (correct decision, but via "unrecognized: for" not rm rules)
- `for i in *; do ls "$i"; done` → ASK (false positive — should be ALLOW)

Tracked in #10 with a proposed fix.

File: src/parse/shell.rs [hunk 3]
```diff
@@ -1,740 +1,804 @@
...
                         continue;
                     }
-                    _ => {
-                        buf.push(c2);
-                        i += 1;
-                    }
+                    let end = effective_end(node).min(child.start_byte());
+                    full.append(
+                        WalkResult::single(sib.start_byte(), end, redir.clone()),
+                        None,
```
[the remark this answers is not in the document]
Comment (butterflysky-ai) [reply src/parse/shell.rs]: Confirmed and fixed in 7d5536b.

The heredoc fast path at line 449 was calling `WalkResult::single(sib.start_byte(), end, redir)` unconditionally — creating a flat text segment from the byte range. For leaf commands (`cat`, `echo`) this is fine, but compound nodes (for/while/if/case) need recursive `walk_ast()` to extract inner commands.

Example: `while true; do shred /dev/sda; done <<EOF | cat` was producing segment `"while true; do shred /dev/sda; done"` → `while` as base command → "unrecognized" → ASK. Should be: unwrap to `shred /dev/sda` → DENY.

Fix adds the same `is_leaf_command()` check that the non-heredoc second pass (line 474) already had. Compound bodies now go through `walk_ast()` with redirection propagation.

Tests added:
- Unit: `compound_heredoc_pipe_unwraps_body` (parser-level segment verification)
- Integration: `while_heredoc_pipe_shred_denies` (end-to-end DENY), `for_heredoc_pipe_ls_allows` (end-to-end ALLOW)

Good catch — this was a real security-relevant gap. Two of three codex reviews on this PR have found legitimate bugs.

**Comment by butterflysky-ai** [comment]
@codex review

File: src/parse/shell.rs [hunk 4]
```diff
@@ -1,740 +1,829 @@
...
-    if i < len && chars[i] == '<' {
-        return None;
+/// Walk `node` for outermost `command_substitution` and `process_substitution`
+/// nodes, appending each to `out`.  Does not recurse into found substitutions;
+/// the eval layer handles nested evaluation.
+fn collect_substitutions(node: Node, source: &[u8], out: &mut Vec<SubstSpan>) {
+    if matches!(node.kind(), "command_substitution" | "process_substitution") {
+        let full = node.utf8_text(source).unwrap_or("");
+        let inner = strip_subst_delimiters(full);
+        if !inner.is_empty() {
```
[the remark this answers is not in the document]
Comment (butterflysky-ai) [reply src/parse/shell.rs]: Not confirmed — this is already handled correctly.

Test:
```
$ printf '{"tool":{"name":"Bash","input":{"command":"cat <<EOF\n\$(shred /dev/sda)\nEOF"}}}' | cargo run --quiet
```
```json
{
  "permissionDecision": "deny",
  "permissionDecisionReason": "compound command (1 substitution(s)):\n  subst[$(shred /dev/sda)] -> DENY: blocked command: shred\n  [cat <<EOF] -> ALLOW: allowed: cat"
}
```

tree-sitter-bash *does* produce `command_substitution` nodes inside unquoted heredoc bodies, and `collect_substitutions` walks the full AST to find them. The substitution is extracted, recursively evaluated, and `shred` triggers DENY.

For quoted heredocs (`cat <<'EOF'`), bash suppresses expansion — no `$()` is evaluated at runtime — so there's nothing to catch. tree-sitter-bash reflects this by not producing `command_substitution` nodes inside quoted heredoc bodies, which is the correct behavior.

[the remark this answers is not in the document]
Comment (butterflysky) [reply src/parse/shell.rs]: explicit tests for command substitution inside quoted and unquoted heredocs added
